JSTOR content banner layout

lyzadanger opened the pull request:
This adds layout for the (JSTOR) content banner.

Wide screens:

<img width="1567" alt="image" src="https://user-images.githubusercontent.com/439947/181568558-1184e15a-8f4f-4641-a95e-a53da075a08a.png">

Narrower screens:

<img width="1174" alt="image" src="https://user-images.githubusercontent.com/439947/181568704-f3f998b1-f74b-4bdd-a074-044d3f70bacc.png">

lyzadanger commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
  * A banner that displays information about the current document and the entity
  * that is providing access to it (eg. JSTOR).
  *
+ * layout columns:
```
The layout in this component is relatively complex. I've attempted to comment thoroughly but let me know if anything feels confusing.

lyzadanger commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+        data-testid="content-container-info"
+      >
+        <span className="font-semibold" title={info.item.containerTitle}>
+          {info.item.containerTitle}
```
Can we assume the presence of `info.item.containerTitle` or should we be more conservative?

robertknight replied:
I'm not sure. There are a bunch of test cases of various document types at https://gist.github.com/robertknight/a2b045b717d6c7dda18879df91c16e90 that I can check.

lyzadanger commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+        >
+          <LinkUnstyled
+            title={`Open "${info.item.title}" in a new window`}
+            href="http://www.example.org"
```
Do you want me to fake this out differently?

lyzadanger commented on src/annotator/integrations/pdf.js:
```diff
@@ -278,10 +278,7 @@ export class PDFIntegration extends TinyEmitter {
-            info={this._bannerState.contentInfo}
-            onClose={() => this._updateBannerState({ contentInfo: null })}
-          />
+          <ContentInfoBanner info={this._bannerState.contentInfo} />
```
No more close button...

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+        />
+        {info.links?.previousItem && (
+          <>
+            <Link
```
All links in this banner should have `target="_blank"` on them at present, since we don't want to navigate the iframe where the user is annotating in the LMS context.

lyzadanger replied:
Derp. Oversight. Will add on next update.

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+              data-testid="content-previous-link"
+            >
+              <CaretLeftIcon className="w-em h-em" />
+              <span>Previous</span>
```
The computed font size of these links is 12px which is pretty small. In the sidebar the smallest font size we use is 13px (at least, that appears to be the case to me). Perhaps we should increase the smallest font size in the annotator to match?

lyzadanger replied:
I agree with you wholeheartedly but I'm not sure this is the time to open this particular Pandora's Box.

robertknight replied:
OK. Perhaps worth filing an issue about though?

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+          <>
+            <Link
+              classes="flex gap-x-1 items-center text-annotator-sm whitespace-nowrap"
+              title={`Open previous item in a new window`}
```
I don't think we need the "in a new window" suffix. In fact I would avoid using a title at all, since the label will be more specific (eg. "Previous chapter", "Previous article"). If we put `target="_blank"` on the link then the browser can provide whatever hints it normally provides to indicate that the link will open in a new tab.

lyzadanger replied:
@robertknight  Oh, note, we don't currently have data for what type of content it is. So if we want to add "chapter" or "article", we'll need that.

lyzadanger replied:
Note that I am still keeping a `title` here because we need some way to show the full title of a potentially-truncated item on hover (disclosure). But I am removing the other language.

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+              data-testid="content-previous-link"
+            >
+              <CaretLeftIcon className="w-em h-em" />
+              <span>Previous</span>
```
Does the `<span>` here do anything, as opposed to just having the text on its own?

robertknight replied:
Also at one point we discussed having content-dependent link titles (eg. "Previous chapter", "Previous article") etc. Is that something we don't need to worry about any more given the current design?

lyzadanger replied:
Holdover from when I had the item type (article or chapter) as part of the link text. I can remove.

lyzadanger replied:
> Also at one point we discussed having content-dependent link titles (eg. "Previous chapter", "Previous article") etc. Is that something we don't need to worry about any more given the current design?

We'd need that metadata (content item "type") available in the config...

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+        'grid items-center',
+        // Two columns in narrower viewports; three in wider
+        'grid-cols-[100px_minmax(0,auto)] gap-x-4',
+        '2xl:grid-cols-[100px_minmax(0,auto)_minmax(0,auto)] 2xl:gap-x-3'
```
On my 13" MacBook Pro, which has a CSS screen size of 1440x900px, I never see the three-column layout. According to [Tailwind's docs](https://tailwindcss.com/docs/screens#using-custom-screen-names), 2xl is 1536px.

robertknight replied:
Would be interesting to see what this looks like with the longest article and journal titles we can find. I'm afraid I don't have an example to hand.

lyzadanger replied:
Well, yes, because :facepalm: I didn't move over some breakpoint definitions from my working branch into the tailwind config. To extend the continuum of our current breakpoints, I intended to add:

```
        xl: '1024px',
        '2xl': '1220px',
        '3xl': '1400px',
```

So, `2xl` at `1220px`. This is, as you might observe, quite conservative still, but necessary because we have to account for the possibility of A) the sidebar being open and B) not being in side-by-side mode (sidebar obscures part of the content).

lyzadanger replied:
> Would be interesting to see what this looks like with the longest article and journal titles we can find. I'm afraid I don't have an example to hand.

I did quite a bit of that when prototyping this design...

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+        data-testid="content-container-info"
+      >
+        <span className="font-semibold" title={info.item.containerTitle}>
+          {info.item.containerTitle}
```
I don't think we explicitly discussed whether the container label should be a link, or whether we even have a link that we could use for that. So there is nothing to do for now. It might come up in future though.

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+          '2xl:block 2xl:whitespace-nowrap 2xl:overflow-hidden 2xl:text-ellipsis'
         )}
+        data-testid="content-container-info"
+      >
```
This `<div>` only has one `<span>` child at present. Would it make sense to merge the parent/child or is there a reason that won't work?

lyzadanger replied:
I'll take a look.

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,108 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+            <Link
+              title={`Open next item in a new window`}
+              classes="flex gap-x-1 items-center text-annotator-sm whitespace-nowrap"
+              href="https://jstor.org"
```
The `href` here should be `info.links.nextItem`.

lyzadanger replied:
Yes, it should. Boy, I'm really on fire for quality control today, eh? 😬 Thanks for your patience.

lyzadanger commented:
@robertknight Pushed updates that fix some idiocies here, including minor feedback you left and:

* Fixed breakpoints
* Added `target="_blank"` to all links. Ensured I was linking the correct thing in general

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,111 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
  * A banner that displays information about the current document and the entity
  * that is providing access to it (eg. JSTOR).
  *
+ * layout columns:
```
```suggestion
 * Layout columns:
```

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,111 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+    >
+      <div data-testid="content-logo">
+        {info.logo && (
+          <Link href="https://jstor.org" target="_blank">
```
```suggestion
          <Link href={info.logo.link} target="_blank">
```

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,111 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+          'font-semibold'
         )}
+        data-testid="content-container-info"
+        title={info.item?.containerTitle}
```
`item` is a non-nullable property, so you don't need to test for it. `containerTitle` _may_ however be an empty string or missing. Same with the line below.

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,111 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+          )}
+        >
+          <LinkUnstyled
+            title={info.item?.title}
```
```suggestion
            title={info.item.title}
```

`info.item` is non-nullable. The `title` property _should_ always be set. If the work is inherently untitled then some suitable placeholder will have been generated by either the LMS app or the JSTOR API.

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,39 +15,111 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+          </LinkUnstyled>
+        </div>
+
+        {info.links?.nextItem && (
```
The `links` property is also non-nullable. The links within that object are optional however (eg. first and last items in a collection or a standalone item won't have such links).

robertknight reviewed (commented):
There are some null checks for `info.item` and `info.links` which are not needed. Also the JSTOR logo should point to the URL in `info.logo.link`. Otherwise looks good to go once tests are updated.

lyzadanger commented:
@robertknight I've removed the link to the current item for now, until that field is available in the content metadata.

robertknight commented:
Hmm... I'm not seeing the title being elided. Tested in Safari and Chrome.

<img width="1177" alt="Banner eliding" src="https://user-images.githubusercontent.com/2458/182390960-cee743af-8f04-4a83-aea1-117233ab9244.png">

lyzadanger commented:
> Hmm... I'm not seeing the title being elided. Tested in Safari and Chrome.

On it. This broke when I changed this from a link to a `<div>`.

robertknight commented on src/annotator/components/ContentInfoBanner.js:
```diff
@@ -8,40 +14,111 @@ import { LabeledButton, Link } from '@hypothesis/frontend-shared';
+            'min-w-0 whitespace-nowrap overflow-hidden text-ellipsis shrink font-medium'
+          )}
+        >
+          <span title={info.item.title} data-testid="content-item-title">
```
The `title` attribute isn't needed here as long as this is a link.

lyzadanger replied:
But it's not a link? :)

robertknight replied:
Sorry, I meant to write "is not a link". Please disregard this comment, it is wrong anyway as the function of `title` is to make the full title visible in a tooltip.

robertknight commented on src/annotator/components/test/ContentInfoBanner-test.js:
```diff
@@ -19,26 +20,76 @@ const contentInfo = {
+    const logoImg = wrapper.find('img[data-testid="logo-image"]');
+
+    assert.equal(logoLink.prop('href'), contentInfo.logo.link);
+    assert.equal(logoImg.prop('src'), 'https://www.jstorg.org/logo.svg');
```
```suggestion
    assert.equal(logoImg.prop('src'), 'https://www.jstor.org/logo.svg');
```

Not that it matters, but I assume this was a typo.

robertknight commented on src/annotator/components/test/ContentInfoBanner-test.js:
```diff
@@ -19,26 +20,76 @@ const contentInfo = {
+      );
+    });
+
+    it('does not display previous link if unavailable', () => {
```
Could do with a similar test that omits the "next" link too.
